### PR TITLE
[Nikhil] - Sage - Build an On-Chain Puzzle Game on Celo

### PR DESCRIPTION
**Pull request by nikbhintade**
### Thanks for submitting this tutorial for Celo Sage

Before you hit "Create Pull Request", make sure you have read the [Tutorial Contribution Guidelines](https://docs.celo.org/community/celo-sage/contribution-guidelines) and reviewed the [Sample Tutorial Structure](https://docs.celo.org/community/celo-sage/tutorial-structure).

Please confirm the following before submitting your tutorial for review -

- [x] My tutorial follows the [Tutorial Contribution Guidelines](https://docs.celo.org/community/celo-sage/contribution-guidelines).
- [x] The title of the Pull Request matches the title of my tutorial.
- [x] I've followed the tutorial template structure.
- [x] I've proofread the tutorial for spelling and grammar mistakes.
- [x] I've thoroughly explained the code, with inline code comments and text explanations.
- [x] I've run a linter & prettifier to format my code blocks.
- [x] I've run the tutorial to test it and the code works.

> NOTE: We cannot review your Pull Request if it does not meet the standard outlined above.
> Reviewers - @joenyzio & @viral-sangani

### CODE REVIEW

**Comment by nikbhintade** [comment]
@viral-sangani I have not put the preview link in `users.tsx` file as I wanted to change title of the tutorial from `Build a Guessing Game on Celo` to `Build an On-Chain Puzzle Game on Celo`. As this is much more suitable title for this tutorial. I have mentioned this on [Trello](https://trello.com/c/FjPlNbgi/138-build-a-guessing-game-on-celo) too.

If you and Joe think that is a suitable title, we can change the text on preview image and then I can update it. Thanks!

**Comment by viral-sangani** [comment]
Hey Nikhil, I have asked Joe to look into that. 
Thanks

**Comment by viral-sangani** [comment]
Hey @nikbhintade Since the title is changed, can you update `users.tsx`?

**Comment by nikbhintade** [comment]
@viral-sangani Sorry for the delay. I have updated the `user.tsx`.

**Comment by nikbhintade** [comment]
@viral-sangani Just wanted to know if there any other changes here.

**Comment by nikbhintade** [comment]
@viral-sangani I wanted to check if there are any other changes remaining?

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+title: Build an On-Chain Puzzle Game on Celo
+description: Learn how to build a On-Chain Puzzle Game on Celo using Solidity and Hardhat.
+authors:
+  - name: Nikhil Bhintade
+tags: [celosage, intermediate, hardhat, solidity]
+hide_table_of_contents: true
+slug: "/tutorials/build-an-on-chain-puzzle-game-on-celo"
+---
+
+# Introduction
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:11]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+In this tutorial, we will be building a puzzle game where the solution to each game will be hashed and stored on the blockchain. This tutorial is designed to teach you about Solidity development, testing, and deployment, as well as how hashing works. Let's get started!
+
+# Prerequisites
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:15]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+To follow this tutorial, it is recommended that you have a basic understanding of Solidity and JavaScript. If you are already familiar with the Hardhat, that will be helpful, but if not, don't worry—we will cover a few things about Hardhat in this tutorial as well. Having some familiarity with these technologies will make it easier for you to understand and follow along with the tutorial, but even if you are new to Solidity and JavaScript, you should still be able to follow along and learn a lot from this tutorial.
+
+# Requirements
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:19]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+To use the `solidity-coverage` plugin, we will need to modify the `hardhat.config.js` file by adding the following line:
+
+```jsx
+require('solidity-covergae')
+```
+
+Once the hardhat project has been set up, you will need to remove all the files from the `contracts` and `test` directories. Then, you can create a new Solidity file called `Puzzle.sol` in the `contracts` directory, and a new test file called `puzzle.test.js` in the `test` directory. These files will be used to develop and test your puzzle game smart contract. After creating these files, you should be ready to start coding your smart contract.
+
+# Writing the Puzzle Game Smart Contract
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:67]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+            _puzzle.solved = true;
+            emit solved(msg.sender, index);
+            return "RIGHT_ANSWER";
+        } else {
+            return "WRONG_ANSWER";
+        }
+    }
+```
+
+# Testing the Puzzle Smart Contract
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:141]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+```bash
+npx hardhat test
+```
+
+The test results should be displayed in the terminal should look similar to following:
+
+![Output of test command on terminal](./images/test-output.png)
+
+# Coverage of Smart Contract Tests
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:264]: **You should add another # hashtag here, to make it a double hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+npx hardhat coverage
+```
+
+You should see report line this in the terminal:
+
+![Output of coverage command on terminal](./images/coverage-output.png)
+
+This command will run all of the tests in your project and generate a coverage report that shows which lines of code were covered by the tests. The report will also be saved in the **`coverage`** folder of your project and can be viewed by opening the **`index.html`** file in a web browser.
+
+# Conclusion
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:278]: **You should add a double # hashtag here, to make it a triple hashtag**

Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:278]: **You should add a double # hashtag here, to make it a triple hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+In this tutorial, we learned about hashing and how it can be used in smart contracts. We implemented this concept using the `keccak256` function in our puzzle smart contract. We then wrote tests to ensure that the contract was functioning correctly. Overall, this tutorial provided a great introduction to hashing and its practical application in Solidity.
+
+# Next Steps
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:282]: **You should add a double # hashtag here, to make it a triple hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+In this tutorial, we not only learned about the concept of hashing in general, but we also implemented it using the `keccak256` function in our puzzle game smart contract. However, there are other ways to implement hashing in smart contracts, such as using the `sha256` function or creating a custom hashing function. It would be interesting to explore these other options and compare their benefits and drawbacks.
+
+Additionally, although we have tested our puzzle game contract and achieved 100% test coverage, there may be other tests that could be added to further improve the robustness of our tests. Some ideas could include testing edge cases, such as trying to solve a puzzle with a very long or very short answer, or testing the contract's behavior when there are a large number of puzzles in the mapping. By adding these additional tests, we can increase our confidence in the contract's correctness and ensure that it is working as expected.
+
+# About the Author
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:288]: **You should add a double # hashtag here, to make it a triple hashtag**

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+Nikhil Bhintade is the author of the article. Nikhil is currently working as a product manager/technical writer and is actively seeking new opportunities in this field. He enjoys writing about cutting-edge technology and the things he is learning. You can see his most recent work on [GitHub](https://github.com/nikbhintade). You can also connect with Nikhil on [Linkedin](https://linkedin.com/in/nikbhintade). If you have any questions or ideas, feel free to send Nikhil a message on LinkedIn.
+
+# References
```
Comment (Julius170) [inline blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:292]: **You should add a double # hashtag here, to make it a triple hashtag**

File: src/data-tutorials/users.tsx [hunk 11]
```diff
@@ -1398,7 +1405,7 @@ const Users: User[] = [
     title: 'How to Create your own DAO on Celo',
     description: 'Build your own DAO with on-chain governance',
     preview: require('./showcase/intermediate/how-to-create-your-own-dao-on-celo.png'),
-    website: 'blog/tutorials/how-to-create-your-own-dao-on-celo',
+    website: '/blog/tutorials/how-to-create-your-own-dao-on-celo',
```
Comment (Julius170) [inline src/data-tutorials/users.tsx:1408]: Arigato

**Review by Julius170 — COMMENTED** [review]
@joenyzio @viral-sangani 
This PR should be good to go after @nikbhintade addresses the reviews I've made

File: blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md (hunk 1 above)
Comment (nikbhintade) [reply blog/01-15-23-Build-On-Chain-Puzzle-Game/01-15-23-Build-An-On-Chain-Puzzle-Game-On-Celo.md:11]: I think I am following previous guideline in this so that is why all the headings are h1 instead of h2. I'll change it soon.

**Comment by nikbhintade** [comment]
@viral-sangani If this is PR is good to go can it be merged?

**Comment by nikbhintade** [comment]
Hi @viral-sangani, I wanted to know if there is any problem with this PR as it is not merged yet. Please let me know if there is any issue. Thanks!

**Review by viral-sangani — APPROVED** [review]

**Comment by viral-sangani** [comment]
Hey Nikhil, So sorry this PR was burried in other PRs that I missed it. Merging it right now.
